src/hal/lpc17xx/pwm.rs

```diff
@@ -0,0 +1,300 @@
+#[derive(Clone, Copy)]
+pub struct PWM {
+  channel: PWMChannel,
+  period_us: u32,
+  pulsewidth_us: u32,
+}
+
+impl PWM {
+
+  /// Create a new PWM Output on the provided channel
+  pub fn new(channel: PWMChannel) -> PWM {
+    PWM1Clock.enable();
+    PWM1Clock.set_divisor(PWM_CLOCK_DIVISOR);
+    reg::PWM1.pr.set_value(0);  // no prescaler
+
+    // single PWM mode (reset TC on match 0 for Ch0)
+    reg::PWM1.mcr.set_pwmmr0r(true);
+
+    // enable PWM output on this channel
+    match channel {
+      CHANNEL0 => { unsafe { abort() } },  // CHANNEL0 reserved for internal use
+      CHANNEL1 => { reg::PWM1.pcr.set_pwmena1(true); },
+      CHANNEL2 => { reg::PWM1.pcr.set_pwmena2(true); },
+      CHANNEL3 => { reg::PWM1.pcr.set_pwmena3(true); },
+      CHANNEL4 => { reg::PWM1.pcr.set_pwmena4(true); },
+      CHANNEL5 => { reg::PWM1.pcr.set_pwmena5(true); },
+      CHANNEL6 => { reg::PWM1.pcr.set_pwmena6(true); },
+    };
+
+    let pwm = PWM {
+      channel: channel,
+      period_us: 20_000,  // 20ms is pretty common
+      pulsewidth_us: 0,
+    };
+
+    pwm.update_period();
+    pwm.update_pulsewidth();
+    pwm
+  }
+
+  /// Update the PWM Signal based on the current state
+  fn update_period(&self) {
+    // Put the counter into reset and disable the counter
+    reg::PWM1.tcr
+      .set_ctr_en(reg::PWM1_tcr_ctr_en::DISABLED)
+      .set_ctr_reset(reg::PWM1_tcr_ctr_reset::RESET)
+      .set_pwm_enable(reg::PWM1_tcr_pwm_enable::DISABLED);
+
+    // setup match register to ticks per period on CH0
+    CHANNEL0.set_mr(pwm_us_to_ticks(self.period_us));
+
+    // TODO: recalculate other registers based on the new period?
+
+    // set the channel latch to update CH0 and CHN
+    reg::PWM1.ler.set_value(1 << CHANNEL0 as u32);
+
+    // enable counter and pwm; clear reset
+    reg::PWM1.tcr
+      .set_ctr_en(reg::PWM1_tcr_ctr_en::ENABLED)
+      .set_ctr_reset(reg::PWM1_tcr_ctr_reset::CLEAR_RESET)
+      .set_pwm_enable(reg::PWM1_tcr_pwm_enable::ENABLED);
+  }
+
+  fn update_pulsewidth(&self) {
+    // TODO: mbed increases ticks by one if ticks for pulsewidth and period
+    //   are equal.  In testing, however, this caused problems.
+    let pulsewidth_ticks: u32 = pwm_us_to_ticks(self.pulsewidth_us);
+    self.channel.set_mr(pulsewidth_ticks);
+    reg::PWM1.ler.set_value(1 << self.channel as u32);
+  }
+
+}
+
+/// Implementation of Generic PWMOutput trait for LPC17xx
+impl ::hal::pwm::PWMOutput for PWM {
+  fn set_period_us(&mut self, period_us: u32) {
+    self.period_us = period_us;
+    self.update_period();
+  }
+
+  fn get_period_us(&self) -> u32 {
+    self.period_us
+  }
+
+  fn set_pulsewidth_us(&mut self, pulsewidth_us: u32) {
+    self.pulsewidth_us = pulsewidth_us;
+    self.update_pulsewidth();
+  }
+
+  fn get_pulsewidth_us(&self) -> u32 {
+    self.pulsewidth_us
+  }
+}
+
+
+/// LPC17xx PWM Register Definitions (User Manual: 24.6)
+mod reg {
+  use volatile_cell::VolatileCell;
+  use core::ops::Drop;
+
+  ioregs!(PWM1 = {
```

Comment:
Can you please refactor with placement (`@ addr`) and remove the extern and linker def?